source/common/common/matchers.cc

```diff
@@ -3,87 +3,115 @@
 #include "envoy/api/v2/core/base.pb.h"
 
 #include "common/config/metadata.h"
-#include "common/protobuf/protobuf.h"
 
 #include "absl/strings/match.h"
 
 namespace Envoy {
 namespace Matchers {
 
-bool DoubleMatcher::match(double value) const {
+MatcherPtr ValueMatcher::create(const envoy::type::matcher::ValueMatcher& v) {
+  switch (v.match_pattern_case()) {
+  case envoy::type::matcher::ValueMatcher::kNullMatch:
+    return std::make_shared<const NullMatcher>();
+  case envoy::type::matcher::ValueMatcher::kDoubleMatch:
+    return std::make_shared<const DoubleMatcher>(v.double_match());
+  case envoy::type::matcher::ValueMatcher::kStringMatch:
+    return std::make_shared<const StringMatcher>(v.string_match());
+  case envoy::type::matcher::ValueMatcher::kBoolMatch:
+    return std::make_shared<const BoolMatcher>(v.bool_match());
+  case envoy::type::matcher::ValueMatcher::kPresentMatch:
+    return std::make_shared<const PresentMatcher>(v.present_match());
+  case envoy::type::matcher::ValueMatcher::kListMatch:
+    return std::make_shared<const ListMatcher>(v.list_match());
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+bool NullMatcher::match(const ProtobufWkt::Value& value) const {
+  return value.kind_case() == ProtobufWkt::Value::kNullValue;
+}
+
+bool BoolMatcher::match(const ProtobufWkt::Value& value) const {
+  return value.kind_case() == ProtobufWkt::Value::kBoolValue && matcher_ == value.bool_value();
+}
+
+bool PresentMatcher::match(const ProtobufWkt::Value& value) const {
+  return matcher_ && value.kind_case() != ProtobufWkt::Value::KIND_NOT_SET;
+}
+
+bool DoubleMatcher::match(const ProtobufWkt::Value& value) const {
+  if (value.kind_case() != ProtobufWkt::Value::kNumberValue) {
+    return false;
+  }
+
+  double v = value.number_value();
   switch (matcher_.match_pattern_case()) {
   case envoy::type::matcher::DoubleMatcher::kRange:
-    return matcher_.range().start() <= value && value < matcher_.range().end();
+    return matcher_.range().start() <= v && v < matcher_.range().end();
   case envoy::type::matcher::DoubleMatcher::kExact:
-    return matcher_.exact() == value;
+    return matcher_.exact() == v;
   default:
     NOT_REACHED_GCOVR_EXCL_LINE;
   };
 }
 
-bool StringMatcher::match(const std::string& value) const {
+bool StringMatcher::match(const ProtobufWkt::Value& value) const {
```

Comment:
Previously, the StringMatcher accepts a simple string so it could be used in anywhere with a string value, now it only can be only used with Value? is this expected?